### PR TITLE
All: Some minors fixes

### PR DESCRIPTION
**Pull request by khancyr**
About the use of fence in copter, avoid use should be restricted to FENCE define too ?

### CODE REVIEW

File: libraries/AP_InertialSensor/AP_InertialSensor_MPU6000.cpp [hunk 1]
```diff
@@ -595,7 +595,7 @@ bool AP_InertialSensor_MPU6000::_accumulate_fast_sampling(uint8_t *samples, uint
     }
 
     if (ret) {
-        float temp = (tsum/n_samples)/340.0 + 36.53;
```
Comment (magicrub) [inline libraries/AP_InertialSensor/AP_InertialSensor_MPU6000.cpp]: @rmackay9 I thought I heard you say we are somehow declaring these constants as floats instead of doubles in the compiler somehow?

File: libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp [hunk 2]
```diff
@@ -438,7 +438,7 @@ bool AP_InertialSensor_MPU9250::_accumulate(uint8_t *samples, uint8_t n_samples)
             _fifo_reset();
             return false;
         }
-        float temp = t2/340 + 36.53;
```
Comment (magicrub) [inline libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp]: there's been a bit of back and forth about casting using C++ method instead of C. Do we have an official stance on that? If so, we should convert the whole project to C++

**Comment by magicrub** [comment]
@CraigElder This LGTM although before we merge I'd like to figure out our official stance on C++ casting which we'll probably talk about at the next dev call.

**Comment by khancyr** [comment]
@magicrub As we use c++ I am in favor of c++ cast that have numerous advantages over old C cast. But I know that @lucasdemarchi dislike the syntax ^^ . One major pb rarely stated that I see is that most static analyzers will produce a warning with old C cast nowadays, and with c++ cast

**Comment by khancyr** [comment]
Also need @rmackay9 thought about first correction. I made it simple but I think the way it done currently isn't safe

**Comment by lucasdemarchi** [comment]
> @magicrub As we use c++ I am in favor of c++ cast that have numerous advantages over old C cast. But I know that @lucasdemarchi dislike the syntax ^^

I find it too verbose, but I don't mind using it actually.

**Comment by rmackay9** [comment]
I must say that I also find the c++ casting syntax verbose and hard to look at and ever so slightly more complex than it really is.  Perhaps i'll get use to it.  I leave it to the better judgement of the better coders in the group - i.e. @tridge, @lucasdemarchi and @OXINARF to decide.

I'm totally happy with the Copter related changes above going in so merge away!

File: libraries/AP_InertialSensor/AP_InertialSensor_MPU6000.cpp (hunk 1 above)
Comment (OXINARF) [inline libraries/AP_InertialSensor/AP_InertialSensor_MPU6000.cpp]: We do for PX4 boards. But the issue here is different: tsum and n_samples are both of integer type so this will perform an integer division and so we can lose a fractional part of the division.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp [hunk 3]
```diff
@@ -508,31 +508,31 @@ void AP_MotorsHeli_Single::write_aux(float servo_out)
 // servo_test - move servos through full range of movement
 void AP_MotorsHeli_Single::servo_test()
 {
-    _servo_test_cycle_time += 1.0f / _loop_rate;
+    _servo_test_cycle_time += 1.0f / static_cast<float>(_loop_rate);
```
Comment (OXINARF) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp]: This one looks not needed...?

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp [hunk 4]
```diff
@@ -508,31 +508,31 @@ void AP_MotorsHeli_Single::write_aux(float servo_out)
...
 
     if ((_servo_test_cycle_time >= 0.0f && _servo_test_cycle_time < 0.5f)||                                   // Tilt swash back
         (_servo_test_cycle_time >= 6.0f && _servo_test_cycle_time < 6.5f)){
-        _pitch_test += (1.0f / (_loop_rate/2));
-        _oscillate_angle += 8 * M_PI / _loop_rate;
+        _pitch_test += (1.0f / (static_cast<float>(_loop_rate)/2));
+        _oscillate_angle += 8 * M_PI / static_cast<float>(_loop_rate);
```
Comment (OXINARF) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp]: The second looks not needed, M_PI is a float.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp [hunk 5]
```diff
@@ -508,31 +508,31 @@ void AP_MotorsHeli_Single::write_aux(float servo_out)
...
         _yaw_test = 0.5f * sinf(_oscillate_angle);
     } else if ((_servo_test_cycle_time >= 0.5f && _servo_test_cycle_time < 4.5f)||                            // Roll swash around
                (_servo_test_cycle_time >= 6.5f && _servo_test_cycle_time < 10.5f)){
-        _oscillate_angle += M_PI / (2 * _loop_rate);
+        _oscillate_angle += M_PI / (2 * static_cast<float>(_loop_rate));
```
Comment (OXINARF) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp]: Not needed. The multiplication is always an integer and M_PI is a float.

File: libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp [hunk 6]
```diff
@@ -438,7 +438,7 @@ bool AP_InertialSensor_MPU9250::_accumulate(uint8_t *samples, uint8_t n_samples)
...
+        float temp = static_cast<float>(t2)/340 + 36.53;
```
Comment (OXINARF) [inline libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp]: I would prefer to write 340.0f than the static cast.

**Comment by OXINARF** [comment]
As Lucas and Randy I also find it too verbose but I would like to see our code migrate to C++ as much as possible. Let's discuss it in the next dev call and make a rule for it - then as changes come in we look that it follows the rule, no need to change all code immediately.

File: libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp (hunk 6 above)
Comment (WickedShell) [inline libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp]: @OXINARF I'd agree with that. I find the C++ one to be very verbose for the purpose.

**Comment by khancyr** [comment]
ok rebase and correct !

File: libraries/AP_Math/matrix_alg.cpp [hunk 7]
```diff
@@ -75,7 +75,7 @@ static void mat_pivot(float* A, float* pivot, uint8_t n)
 {
     for(uint8_t i = 0;i<n;i++){
         for(uint8_t j=0;j<n;j++) {
-            pivot[i*n+j] = (i==j);
+            pivot[i*n+j] = static_cast<float>(i==j);
```
Comment (lucasdemarchi) [inline libraries/AP_Math/matrix_alg.cpp]: It would be faster to do something like:

```
memset(pivot, 0, sizeof(float) * n);
for (uint8_t i = 0; i < n; i++) {
    pivot[i*n + i] = 1.f;
}
```

Comment (lucasdemarchi) [inline libraries/AP_Math/matrix_alg.cpp]: But... this change is out of scope for the patch, so that's fine.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp [hunk 8]
```diff
@@ -512,27 +512,27 @@ void AP_MotorsHeli_Single::servo_test()
 
     if ((_servo_test_cycle_time >= 0.0f && _servo_test_cycle_time < 0.5f)||                                   // Tilt swash back
         (_servo_test_cycle_time >= 6.0f && _servo_test_cycle_time < 6.5f)){
-        _pitch_test += (1.0f / (_loop_rate/2));
-        _oscillate_angle += 8 * M_PI / _loop_rate;
+        _pitch_test += (1.0f / (_loop_rate / 2.0f));
+        _oscillate_angle += 8.0f * M_PI / _loop_rate;
```
Comment (OXINARF) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp]: Why the 8.0f? Like I said before, M_PI is a float, 8 should be cast to float.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp [hunk 9]
```diff
@@ -512,27 +512,27 @@ void AP_MotorsHeli_Single::servo_test()
...
         _yaw_test = 0.5f * sinf(_oscillate_angle);
     } else if ((_servo_test_cycle_time >= 0.5f && _servo_test_cycle_time < 4.5f)||                            // Roll swash around
                (_servo_test_cycle_time >= 6.5f && _servo_test_cycle_time < 10.5f)){
-        _oscillate_angle += M_PI / (2 * _loop_rate);
+        _oscillate_angle += M_PI / (2.0f * _loop_rate);
```
Comment (OXINARF) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp]: Why? This does a float multiplication instead of an integer one without any benefit.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp [hunk 10]
```diff
@@ -512,27 +512,27 @@ void AP_MotorsHeli_Single::servo_test()
...
         _yaw_test = 0.5f * sinf(_oscillate_angle);
     } else if (_servo_test_cycle_time >= 5.0f && _servo_test_cycle_time < 6.0f){                              // Raise swash to top
         _collective_test += (1.0f / _loop_rate);
-        _oscillate_angle += 2 * M_PI / _loop_rate;
+        _oscillate_angle += 2.0f * M_PI / _loop_rate;
         _yaw_test = sinf(_oscillate_angle);
     } else if (_servo_test_cycle_time >= 11.0f && _servo_test_cycle_time < 12.0f){                            // Lower swash to bottom
         _collective_test -= (1.0f / _loop_rate);
-        _oscillate_angle += 2 * M_PI / _loop_rate;
+        _oscillate_angle += 2.0f * M_PI / _loop_rate;
```
Comment (OXINARF) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp]: All the 3 changes to float above, same question, why when M_PI is a float?

Comment (khancyr) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp]: It will be implicitly cast to float anyway, so I make them float so that is clear that the operation is done on float

**Comment by lucasdemarchi** [comment]
Applied all commits, except the one to AP_Motors.... I agree with @OXINARF that we shouldn't be doing float operations when fixed point is fine.

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp (hunk 10 above)
Comment (OXINARF) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp]: I prefer to have the integer there (we actually have a M_2_PI, or something like that, that could be used), but I'll ask in today's dev call what the preference is.

File: libraries/AP_Math/matrix_alg.cpp (hunk 7 above)
Comment (OXINARF) [inline libraries/AP_Math/matrix_alg.cpp]: Preventing mistakes is great and removing warnings is awesome too! Thank you for your work.

**Comment by khancyr** [comment]
ok done ! Sorry for the trouble!

**Comment by lucasdemarchi** [comment]
@khancyr Great work here. Don't be sorry :)
